File: CSES/Math/Fixed_Length_Walk_Queries.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll INF = (ll)1e18;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, q;
    cin >> n >> m >> q;
    vector<vector<int>> adj(n);
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        --u; --v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    // dis[src][node*2 + parity] = shortest distance with given parity
    vector<vector<ll>> dis(n, vector<ll>(2 * n, INF));

    vector<int> curr, next;
    for (int src = 0; src < n; src++) {
        auto &d = dis[src];
        curr.clear();
        d[src * 2] = 0;
        curr.push_back(src * 2);

        while (!curr.empty()) {
            next.clear();
            for (int u : curr) {
                int node = u / 2;
                ll dist_u = d[u];
                for (int v : adj[node]) {
                    ll nd = dist_u + 1;
                    int idx = v * 2 + (nd % 2);
                    if (nd < d[idx]) {
                        d[idx] = nd;
                        next.push_back(idx);
                    }
                }
            }
            curr.swap(next);
        }
    }

    while (q--) {
        int a, b, x;
        cin >> a >> b >> x;
        --a; --b;
        ll even_d = dis[a][b * 2];
        ll odd_d  = dis[a][b * 2 + 1];
        bool ok = (x >= even_d && (x - even_d) % 2 == 0)
               || (x >= odd_d  && (x - odd_d)  % 2 == 0);
        cout << (ok ? "YES" : "NO") << '\n';
    }

    return 0;
}

File: CSES/Math/Fixed_Length_Walk_Queries.py
import sys
input = sys.stdin.readline

def main():
    n, m, q = map(int, input().split())
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(lambda x: int(x)-1, input().split())
        adj[u].append(v)
        adj[v].append(u)

    INF = 1 << 60
    dis = [[INF] * (2 * n) for _ in range(n)]

    for src in range(n):
        d = dis[src]
        d[src*2] = 0
        stack = [src*2]
        while stack:
            nxt = []
            for u in stack:
                i = u // 2
                for v in adj[i]:
                    dj = d[u] + 1
                    idx = v * 2 + (dj % 2)
                    if dj < d[idx]:
                        d[idx] = dj
                        nxt.append(idx)
            stack = nxt

    for _ in range(q):
        a, b, x = map(lambda x: int(x)-1, input().split())
        x += 1
        e = dis[a][b*2]
        o = dis[a][b*2+1]
        ok = (x >= e and (x - e) % 2 == 0) or (x >= o and (x - o) % 2 == 0)
        print("YES" if ok else "NO")

main()
